### DISCUSSION TITLE
Extracting path using fallible services?

### DISCUSSION THREAD

**Discussion by joehdodd** [tokio-rs/axum#2446] [Q&A] [answered]
### Summary

Following the example here: https://docs.rs/axum/latest/axum/error_handling/index.html#routing-to-fallible-services I've implemented a way that I can use the `?` operator in handlers that make `reqwest` calls in order to return status codes and error messages sent back from external APIs to my client. But I'm not sure how to extract the path (or anything else needed by the handler) with this method. 

Thanks!

```rust
async fn get_person(
    /*
        Is it possible to extract the person_id (and state) here, given my implementation?
     */
) -> Result<Json<serde_json::Value>, reqwest::Error> {
    let request_url = format!("https://swapi.dev/api/people/{}", "1asdf".to_owned());
    let response = reqwest::get(request_url).await?;
    if response.status().is_success() {
        let json = response.json::<serde_json::Value>().await?;
        Ok(Json(json))
    } else {
        Err(response.error_for_status().unwrap_err())
    }
}

#[tokio::main]
async fn main() {
    let client = reqwest::Client::new();

    let cors = CorsLayer::new()
        .allow_methods([Method::GET])
        .allow_origin(Any);

    let faillible_person_service = tower::service_fn(|_req| async {
        let body = get_person().await?;
        Ok::<_, reqwest::Error>(body.into_response())
    });

    let app = Router::new()
        .route("/", get(get_weather))
        .route_service(
            "/person/:person_id",
            HandleError::new(faillible_person_service, handle_reqwest_error),
        )
        .layer(cors)
        .with_state(client);
    let listener = tokio::net::TcpListener::bind("127.0.0.1:1337")
        .await
        .unwrap();
    axum::serve(listener, app).await.unwrap();
}

async fn handle_reqwest_error(err: reqwest::Error) -> Response {
    let status = err.status();
    let status = status.unwrap_or(reqwest::StatusCode::INTERNAL_SERVER_ERROR);
    let status_as_u16 = status.as_u16();
    let axum_status =
        StatusCode::from_u16(status_as_u16).unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
    let res_json = Json(serde_json::json!({
        "error": {
            "message": format!("Something went wrong: {}", err),
        },
    }));
    return (axum_status, res_json).into_response();
}
```

### axum version

0.7.2

**Comment by davidpdrsn** [comment, marked as answer]
You can only run extractors from handler functions called by axum. get_person is just a normal rust function and isn’t called by axum. So you to pass in the right arguments like any other function yourself.

I recommend you just use normal handlers without bothering with services, similarly to https://github.com/tokio-rs/axum/blob/main/examples/reqwest-response/src/main.rs

**Reply by joehdodd** [reply]
Thanks for the reply! Noted.

**Reply by davidpdrsn** [reply]
> The goal is to be able to use the ? operator in order to look at the actual response and status code from the external API call, rather than the response and status code from the reqwest call.

What’s the difference? reqwest returns the status from the HTTP request.

**Reply by joehdodd** [reply]
Yes, sorry about that! I was confused on my end. Marked your original reply as the answer. Thank you!
